web/packages/teleterm/src/services/appUpdater/clientToolsUpdateProvider.ts

```diff
@@ -0,0 +1,125 @@
+ */
+
+import { app } from 'electron';
+import {
+  AppUpdater,
+  DebUpdater,
+  MacUpdater,
+  NsisUpdater,
+  Provider,
+  ResolvedUpdateFileInfo,
+  RpmUpdater,
+  UpdateInfo,
+} from 'electron-updater';
+import { ProviderRuntimeOptions } from 'electron-updater/out/providers/Provider';
+
+const CHECKSUM_FETCH_TIMEOUT = 5_000;
+
+/** Implements electron-updater's `Provider` with client tools updates. */
+export class ClientToolsUpdateProvider extends Provider<UpdateInfo> {
+  constructor(
+    private getClientToolsVersion: ClientToolsVersionGetter,
+    private nativeUpdater: AppUpdater,
+    runtimeOptions: ProviderRuntimeOptions
+  ) {
+    super(runtimeOptions);
+  }
+
+  override async getLatestVersion(): Promise<UpdateInfo> {
+    const clientTools = await this.getClientToolsVersion();
+
+    // If no client tools version is specified, return the current version
+    // to simulate an up-to-date state.
+    if (!clientTools) {
+      return {
+        version: app.getVersion(),
+        releaseDate: '',
+        path: '',
+        sha512: '',
+        files: [],
+      };
+    }
+
+    const { baseUrl, version } = clientTools;
+    const fileUrl = `https://${baseUrl}/${makeDownloadFilename(this.nativeUpdater, version)}`;
+    const sha256 = await fetchChecksum(fileUrl);
+
+    return {
+      version: version,
+      releaseDate: '',
+      path: '',
+      sha512: '',
+      files: [
+        {
+          // Effective only on Windows.
+          isAdminRightsRequired: true,
+          url: fileUrl,
+          // @ts-expect-error sha2 field doesn't exist in the types but is supported.
+          sha2: sha256,
+        },
+      ],
+    };
+  }
+
+  override resolveFiles(updateInfo: UpdateInfo): ResolvedUpdateFileInfo[] {
+    return updateInfo.files.map(fileInfo => ({
+      url: new URL(fileInfo.url),
+      info: fileInfo,
+    }));
+  }
+}
+
+/** Should return undefined when client tools version is not available. */
+export type ClientToolsVersionGetter = () => Promise<
+  | {
+      /** Base URL for downloading Teleport packages. e.g. cdn.teleport.dev. */
+      baseUrl: string;
+      /** Version to download. */
+      version: string;
+    }
+  | undefined
+>;
+
+function makeDownloadFilename(updater: AppUpdater, version: string): string {
+  if (updater instanceof MacUpdater) {
+    return `Teleport Connect-${version}-mac.zip`;
+  }
+  if (updater instanceof NsisUpdater) {
+    return `Teleport Connect Setup-${version}.exe`;
+  }
+  if (updater instanceof RpmUpdater) {
+    return `teleport-connect-${version}.x86_64.rpm`;
+  }
+  if (updater instanceof DebUpdater) {
+    return `teleport-connect_${version}_amd64.deb`;
+  }
+
+  throw new Error(`Unsupported app updater: ${updater?.constructor?.name}`);
+}
+
+async function fetchChecksum(fileUrl: string): Promise<string> {
+  const checksumUrl = `${fileUrl}.sha256`;
+  const checksum = await fetch(checksumUrl, {
```

Comment:
What's the UX when this request fails or times out?

Reply:
The error is propagated through electron-updater and we will display than in `Alert`. 
In case of any error, there will be a "Try again" button.